### PR TITLE
[RFC] Disable syntax folding update in InsertLeave

### PR DESCRIPTION
**Pull request by Shougo**
It fixes the slowness in InsertLeave if you use syntax/expr folding.

### CODE REVIEW

**Comment by justinmk** [comment]
Did you have any updates to this? Or is it ready?

File: src/nvim/fold.c [hunk 1]
```diff
@@ -788,6 +788,22 @@ void foldUpdate(win_T *wp, linenr_T top, linenr_T bot)
...
+// foldUpdateInsert()
+//
+// Update folds for the insert changes in the buffer of a window.
+//
+void foldUpdateInsert(void)
+{
+  if (foldmethodIsManual(curwin)
+      || foldmethodIsSyntax(curwin)
+      || foldmethodIsExpr(curwin)) {
+    return;
```
Comment (justinmk) [inline src/nvim/fold.c]: Why are only these skipped, but not other fold methods?

**Comment by Shougo** [comment]
> Did you have any updates to this? Or is it ready?

It is ready.

File: src/nvim/fold.c (hunk 1 above)
Comment (Shougo) [inline src/nvim/fold.c]: I think them should be updated.
manual:  No need to update
syntax and expr: Auto update is slow.  It should be updated manually like FastFold plugin

**Comment by justinmk** [comment]
Merged

**Comment by Shougo** [comment]
Thank you!

**Comment by cHoco** [comment]
I understand the reasoning behind this change, but how can I update the folds manually now? `zx` and `zX` don't really do what I want.
It would be cool if this change also included a `:foldupdate command` so I can call it when I want the folds updated.

**Comment by justinmk** [comment]
@Shougo What do you think?

**Comment by Shougo** [comment]
@cHoco Please change `foldmethod` value. It resets the folds.

**Comment by justinmk** [comment]
@cHoco Which foldmethod are you using? And why doesn't `zx` work for you?

**Comment by Shougo** [comment]
Oh, I get it.
I don't use zx ever...

**Comment by cHoco** [comment]
@Shougo that works, but it isn't really "user friendly" :/ I still think a command would be worth it.

@justinmk  I'm using syntax. `zx` applys fold level, so some folds that weren't closed get closed, etc....

Anyway this behaviour change should probably be noted in the documentation.

**Comment by Shougo** [comment]
Hm.  Vim already have `foldopen`, `foldclose`, ... commands.
Why no `:foldupdate` command in Vim.

**Comment by Shougo** [comment]
@cHoco Can you ask the question in the issue?

https://github.com/vim/vim/pull/1045

**Comment by justinmk** [comment]
Wait. Why would we add `:foldupdate` after we just disabled auto-update-of-folds? If this PR broke something we should let auto-updating happen after insert-leave.

**Comment by cHoco** [comment]
After https://github.com/neovim/neovim/pull/5299 and https://github.com/neovim/neovim/pull/5351 folds weren't updated in insert mode while typing, but only when leaving it. This PR disabled folds update at insert leave for the slower modes (and manual since it doesn't need updating anyway).

So this PR broke previous behaviour, new text inserted doesn't update folds. Just commented since I think:

1. change should be mentioned in the documentation, users expect folds to autoupdate.
2. add a `foldupdate` command (complementary to `foldopen` etc.. since `zx` also reapplys fold level) to have a friendly way to update folds.

**Comment by justinmk** [comment]
> This PR disabled folds update at insert leave for the slower modes (and manual since it doesn't need updating anyway).

If user has to manually call (or map) `:foldupdate` it seems like we didn't improve anything. So I think we _should_ auto-update folds after insert-leave, for everything except `foldmethod=manual`.

**Comment by Shougo** [comment]
>So I think we should auto-update folds after insert-leave, for everything except foldmethod=manual.

I don't like the auto update.
It is slow even if you edit 4000 lines help file.

**Comment by cHoco** [comment]
> So I think we should auto-update folds after insert-leave, for everything except `foldmethod=manual`.

Well, updating at insert leave feels really bad when editing even moderately big files :( I personally would call `foldupdate` on `BufWrite` or manually.

The best solution would be to optimize `syntax` fold method but last time I took a look most of the time is spent in the regex engine, so I don't think there's a lot that can be done there without moving away from it.

**Comment by Shougo** [comment]
Here.
https://github.com/vim/vim/pull/1045#issuecomment-248122927

**Comment by Shougo** [comment]
Here.
https://github.com/vim/vim/pull/1045#issuecomment-248122927

**Comment by Shougo** [comment]
Here.
https://github.com/vim/vim/pull/1045#issuecomment-248122927

**Comment by justinmk** [comment]
Ok, how about doing it on `CursorHold` and `BufWrite` ?

**Comment by Shougo** [comment]
`CursorHold` is also slow.
But `BufWritePost` timing is not so bad.

**Comment by cHoco** [comment]
Yeah `BufWritePost` seems reasonable 👍

**Comment by Shougo** [comment]
I have fixed the problem in insert mode.
But foldmethod is also slow in normal mode.  Oh, no...
https://github.com/Shougo/neosnippet.vim/issues/356

**Comment by justinmk** [comment]
> zx also reapplys fold level, so some folds that weren't closed get closed, etc....

@choco Does `zizi` help? What about doing another `set foldenable`?

**Comment by choco** [comment]
@justinmk nope, Shougo method above is the only one working

**Comment by justinmk** [comment]
anyone want to send a patch for `:foldupdate` ?

**Comment by choco** [comment]
If there's no rush I can probably send something (since I'm also the one bugging about it :>), but not before next weekend.

**Comment by justinmk** [comment]
@choco Instead of `:foldupdate`, it might be enough to CTRL-L. Could you try that?

**Comment by choco** [comment]
@justinmk Yep, it seems to work indeed. Sorry for not getting back before, been pretty busy last couple of months.

**Comment by choco** [comment]
@justinmk just wanted to point out that I tried to apply the fix in the pull request above and it doesn't work. I still need to press ctrl-L for the folds to update.

**Pull request closed without merging**